commands/find_stemcell_version.go

```diff
@@ -0,0 +1,138 @@
+package commands
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Masterminds/semver"
+	"github.com/pivotal-cf/jhanda"
+	"github.com/pivotal-cf/kiln/fetcher"
+	"github.com/pivotal-cf/kiln/internal/cargo"
+	"gopkg.in/src-d/go-billy.v4/osfs"
+	"log"
+	"regexp"
+	"strings"
+)
+
+const (
+	ErrStemcellOSInfoMustBeValid       = "Stemcell OS information is missing or invalid"
+	ErrStemcellMajorVersionMustBeValid = "Stemcell Major Version is missing or invalid"
+	TanzunetRemotePath                  = "network.pivotal.io"
+)
+
+type FindStemcellVersion struct {
+	outLogger   *log.Logger
+	pivnetService *fetcher.Pivnet
+
+	Options struct {
+		Kilnfile         string  `short:"kf" long:"kilnfile" default:"Kilnfile" description:"path to Kilnfile"`
+		VariablesFiles []string `short:"vf" long:"variables-file" description:"path to variables file"`
+		Variables      []string `short:"vr" long:"variable" description:"variable in key=value format"`
+	}
+}
+
+type stemcellVersionOutput struct {
+	Version    string `json:"version"`
+	Source     string `json:"source"`
+	RemotePath string `json:"remote_path"`
+}
+
+func NewFindStemcellVersion(outLogger *log.Logger, pivnetService *fetcher.Pivnet) FindStemcellVersion {
+	return FindStemcellVersion{
+		outLogger:   outLogger,
+		pivnetService: pivnetService,
+	}
+}
+
+func (cmd FindStemcellVersion) Execute(args []string) error {
+	kilnfile, err := cmd.setup(args)
+
+	if err != nil {
+		return err
+	}
+
+	var productSlug=""
+
+	//Get Stemcell OS and major from Kilnfile
+	if kilnfile.Stemcell.OS == "ubuntu-xenial" {
+		productSlug = "stemcells-ubuntu-xenial"
+	} else if kilnfile.Stemcell.OS == "windows2019" {
+		productSlug = "stemcells-windows-server"
+	}
+
+    if productSlug == "" {
+		return fmt.Errorf(ErrStemcellOSInfoMustBeValid)
+	} else if kilnfile.Stemcell.Version == "" {
+		return fmt.Errorf(ErrStemcellMajorVersionMustBeValid)
+	} else {
+
+		majorVersion, err := ExtractMajorVersion(kilnfile.Stemcell.Version)
+
+		if err != nil {
+			return err
+		}
+
+		//Get stemcell version from pivnet
+		stemcellVersion, err := cmd.pivnetService.StemcellVersion(productSlug, majorVersion)
+
+		if err != nil {
+			return err
+		}
+
+		stemcellVersionJson, _ := json.Marshal(stemcellVersionOutput{
+			Version: stemcellVersion,
+			Source: "Tanzunet",
+			RemotePath: TanzunetRemotePath,
+		})
+		cmd.outLogger.Println(string(stemcellVersionJson))
+	}
+	return nil
+}
+
+func ExtractMajorVersion(version string) (string, error) {
```

Comment:
(this is a really loosly held opinion... feel free to ignore it)

Version constraints from this package may not match what the code below seems to expect. It might make sense to fetch the versions from Pivnet and then see which version matches the constraint. Another option would be to instead of specifying the version have the key specify the version line. For example:

```go
package cargo

type StemcellConstraint struct {
	Slug         string `yaml:"slug"`
	MajorVersion int    `yaml:"version_line"`
}
```

Reply:
Yhea for now the Kilnfile has options to pin a version. May be we can later expand if we want robust semver rules. I assumed for Stemcell we might not have very complex rules and just limited based on the major alone. But good point where we might have to consider for complex version constraints.